File: sdk/python/seocheon/internal/crypto/bip44.py
"""BIP44 key derivation for the Seocheon SDK."""

from __future__ import annotations

from bip_utils import (
    Bip39SeedGenerator,
    Bip39WordsNum,
    Bip44,
    Bip44Changes,
    Bip44Coins,
)

from seocheon.internal.crypto.keys import PrivateKey


def derive_key_from_mnemonic(mnemonic: str) -> PrivateKey:
    """Derive a secp256k1 private key from a BIP39 mnemonic.

    Uses the Cosmos BIP44 path: m/44'/118'/0'/0/0
    """
    seed = Bip39SeedGenerator(mnemonic).Generate()
    bip44_ctx = Bip44.FromSeed(seed, Bip44Coins.COSMOS)
    key = (
        bip44_ctx
        .Purpose()
        .Coin()
        .Account(0)
        .Change(Bip44Changes.CHAIN_EXT)
        .AddressIndex(0)
    )
    priv_key_bytes = key.PrivateKey().Raw().ToBytes()
    return PrivateKey.from_bytes(priv_key_bytes)

File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/internal/crypto/Bip44.kt
package com.seocheon.sdk.internal.crypto

import org.bitcoinj.crypto.HDKeyDerivation
import org.bitcoinj.crypto.MnemonicCode
import org.bitcoinj.crypto.ChildNumber

/**
 * BIP39/BIP44 key derivation for Cosmos-compatible chains.
 * Path: m/44'/118'/0'/0/0
 */
object Bip44 {

    /**
     * Derives a secp256k1 private key from a BIP39 mnemonic
     * using the Cosmos BIP44 path: m/44'/118'/0'/0/0
     */
    fun deriveKeyFromMnemonic(mnemonic: String): PrivateKey {
        val words = mnemonic.trim().split("\\s+".toRegex())
        MnemonicCode.INSTANCE.check(words)

        val seed = MnemonicCode.toSeed(words, "")

        // Create master key
        val masterKey = HDKeyDerivation.createMasterPrivateKey(seed)

        // Derive m/44'/118'/0'/0/0
        val purpose = HDKeyDerivation.deriveChildKey(masterKey, ChildNumber(44, true))
        val coinType = HDKeyDerivation.deriveChildKey(purpose, ChildNumber(118, true))
        val account = HDKeyDerivation.deriveChildKey(coinType, ChildNumber(0, true))
        val change = HDKeyDerivation.deriveChildKey(account, ChildNumber(0, false))
        val addressKey = HDKeyDerivation.deriveChildKey(change, ChildNumber(0, false))

        return PrivateKey.fromBytes(addressKey.privKeyBytes)
    }

    /**
     * Validates a BIP39 mnemonic phrase.
     */
    fun isValidMnemonic(mnemonic: String): Boolean {
        return try {
            val words = mnemonic.trim().split("\\s+".toRegex())
            MnemonicCode.INSTANCE.check(words)
            true
        } catch (_: Exception) {
            false
        }
    }
}
